### PR TITLE
Create secure API to retrieve list of emails associated with a form

### LINKED ISSUE #492: Create secure API to retrieve list of emails associated with a form
As an admin user I would like to request all emails associated with a form so that I may see who has access to its data 

Acceptance criteria:

- [x] An API route in the NextJS App that accepts `GET` requests. 
- [x] The API route should be in the following file `pages/api/id/[form]/owners.js`
- [x] The API route should only allow requests with valid admin user sessions from next auth. Otherwise a 403 status code should be returned 
- [x] The API route should return all the emails associated with the form ID which will be provided in the path of the request. 
- [x] If there is no emails associated with the specific form then an empty ID should be returned 
- [x] If the form ID cannot be found in the database then a 404 status code should be returned
- [x] Adequate unit testing on all levels

### PR DESCRIPTION
**Pull request by falila**
# Summary | Résumé

As an admin user I would like to request all emails associated with a form so that I may see who has access to its data.
close #492 

Insert Image showing after a change was made 
Screen Shot 2021-11-25 at 8 57 49 AM

# Test instructions | Instructions pour tester la modification

1. Open postman UI
2. restart and run your migration
3. Add emails to form_users  table
4. Paste http://localhost:3000/api/id/{firmID}/owners in Postman replace formID by yours.
5. Get the cookie object from Chrome devTools and add it to Postman header.
6. Hit the url and it should return your bearer token.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [ ] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Have you removed unnecessary lines such as `console.log` statements?
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [x] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [ ] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: migrations/migrations/006-form-users.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+const createFormUsersTable = `
+  CREATE TABLE form_users (
+    id SERIAL PRIMARY KEY,
+    template_id INT NOT NULL REFERENCES templates (id),
```
Comment (Moro-Code) [inline migrations/migrations/006-form-users.js:4]: There needs to be a combined unique constraint on template_id and email to ensure emails don't get added more than once a particular form

File: pages/api/id/[form]/owners.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+export async function getEmailListByFormID(
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<void> {
+  const formID = req.query.form as string;
+  if (formID) {
+    //Get emails by formID
+    const resultObject = await executeQuery(
+      dbConnector(),
+      "SELECT email FROM form_users WHERE template_id = ($1)",
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:30]: should have been specified in the acceptance criteria. It's probably prudent to also return the id's and the status

File: pages/api/id/[form]/owners.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      [formID]
+    );
+    type EmailListResponse = { email: string };
+    //Return all emails associated with formID
+    if (resultObject.rowCount > 0)
+      return res.status(200).json(
+        resultObject.rows.map((elem) => {
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:37]: when adding id and active status you can probably just pass `resultObject.rows` here without the map function

File: pages/api/id/[form]/owners.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    //Get emails by formID
+    const resultObject = await executeQuery(
+      dbConnector(),
+      "SELECT email FROM form_users WHERE template_id = ($1)",
+      [formID]
+    );
+    type EmailListResponse = { email: string };
+    //Return all emails associated with formID
+    if (resultObject.rowCount > 0)
+      return res.status(200).json(
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:36]: This confused me a bit and then I realized its a one line if statement. For readability I think it's okay to add the brackets

File: pages/api/id/[form]/owners.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        resultObject.rows.map((elem) => {
+          const { email } = elem as unknown as EmailListResponse;
+          return email;
+        })
+      );
+    //Otherwise a 404 form Not Found
+    return res.status(404).json({ error: "Form Not Found" });
+  }
+  //Could not find formID in the path
+  return res.status(400).json({ error: "Malformed API Request" });
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:46]: Think the error message could be more specific here

File: tests/api/id/[form]/owners.test.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+jest.mock("../../../../lib/integration/dbConnector", () => {
+  const mClient = {
+    connect: jest.fn(),
+    query: jest.fn(),
+    end: jest.fn(),
+  };
+  return jest.fn(() => mClient);
+});
+
+describe("Test Owners : retrieve list of emails API endpoint", () => {
```
Comment (Moro-Code) [inline tests/api/id/[form]/owners.test.js:18]: Well written tests well done

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Few changes but a much better PR thats for sure!

File: migrations/migrations/006-form-users.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    email VARCHAR(80) NOT NULL,
+    temporary_token TEXT,
+    active BOOLEAN NOT NULL DEFAULT TRUE,
+    created_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    UNIQUE(template_id,active)
```
Comment (Moro-Code) [inline migrations/migrations/006-form-users.js:10]: you probably mean template_id and email here right ?

Comment (Moro-Code) [reply migrations/migrations/006-form-users.js:10]: lets also use the `CONSTRAINT constraint_name UNIQUE (uc_col1, uc_col2, ... uc_col_n)` format as it makes it easier for down the line migrations

File: pages/api/id/[form]/owners.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+export async function getEmailListByFormID(
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<void> {
+  const formID = req.query.form as string;
+  if (formID) {
+    //Get emails by formID
+    const resultObject = await executeQuery(
+      dbConnector(),
+      "SELECT id, email FROM form_users WHERE template_id = ($1)",
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:30]: Need active status as I mentioned last time as well

**Review by Moro-Code — CHANGES_REQUESTED** [review]
A mistake with the migration and an overlooked thing from the last review. Almost there

File: migrations/migrations/006-form-users.js (hunk 7 above)
Comment (falila) [reply migrations/migrations/006-form-users.js:10]: CONSTRAINT template_id_email_unique UNIQUE (template_id,email)

File: tests/api/id/[form]/owners.test.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  });
+
+  it("Should return all the emails associated with the form ID.", async () => {
+    const mockSession = {
+      expires: "1",
+      user: { email: "forms@cds.ca", name: "forms user" },
+    };
+    client.getSession.mockReturnValue(mockSession);
+    // Mocking executeQuery to return a list of emails
+    executeQuery.mockReturnValue({
```
Comment (Moro-Code) [inline tests/api/id/[form]/owners.test.js:67]: hmmm I would expect the active here being a boolean value but its minor

**Review by Moro-Code — APPROVED** [review]
Minor comment if you can fix. It's ready to merge though

**Pull request merged**